### ISSUE TITLE
✨ choose between `pip-sync` and `pip install -r`

### ISSUE THREAD

**Issue by juftin** [juftin/hatch-pip-compile#29]
`pip-sync` has been demonstrating some issues with running lockfiles on different python versions: https://github.com/juftin/hatch-pip-compile/actions/runs/7028915350/job/19125657383

Here's an error from testing the plugin on Python 3.8:

```
Traceback (most recent call last):
...
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/_pytest/_code/code.py", line 59, in <module>
    from exceptiongroup import BaseExceptionGroup
ModuleNotFoundError: No module named 'exceptiongroup'
Error: Process completed with exit code 1.
```

Basically the lockfile was generated using 3.11 but the `exceptiongroup` package wasn't included because it's a builtin in 3.11. The plugin uses `pip-sync` which mirrors the environment to the lockfile (installing and uninstalling) so the tests failed. If we were to just run `pip install -r lockfile.txt` instead though the missing `exceptiongroup` dependency would've been picked up and installed (as long as we don't pass the `--no-deps` argument to `pip`).

I can see an argument for wanting to handle this either way. To get around this issue I used a regular `virtual` environment in my testing matrix instead of generating a lockfile for every environment. I woul've preferred to set an option like `pip-compile-installer` to `pip` instad of `pip-sync` instead https://github.com/juftin/hatch-pip-compile/blob/430b2b8af85d6cb436278d8221a202fb874dac90/pyproject.toml#L113-L129

This could also be handled by just doing a `pip install -r` in CI/CD and bypassing this hatch completely.

**Comment by oprypin** [comment]
Yeah I've never used pip-sync.
How does Hatch handle this in virtual environments? Does it uninstall anything?

**Comment by juftin** [comment]
> Yeah I've never used pip-sync. How does Hatch handle this in virtual environments? Does it uninstall anything?

`pip-sync` is the other side of `pip-compile` but I haven't used it much before either. Yeah - it uninstalls everything not on the lockfile. This is also the reason why the plugin has to go through the extra step of reinstalling the project in dev mode, `pip-sync` uninstalls that too.

**Comment by juftin** [comment]
> How does Hatch handle this in virtual environments? Does it uninstall anything?

I don't _think_ so - it just iterates through the `dependencies` attribute  and requests a `pip install` I think...

**Comment by oprypin** [comment]
If so, sounds like pip-sync could be just dropped ¯\\\_(ツ)_/¯

**Comment by juftin** [comment]
Confirmed, `virtual` will leave dependencies behind and not uninstall them.

Yeah, I'll go ahead and make `pip` the default, keeping the ability for `pip-sync`. The plugin is called `hatch-pip-compile` and not `hatch-pip-compile-pip-sync` so I don't think it warrants a breaking change. 

I'm thinking of `pip-compile-installer` as the config option

**Comment by oprypin** [comment]
Meanwhile...
I got some totally bizarre interaction with uninstalling.

https://github.com/oprypin/mkdocs/commit/28e19f34ce7699db46040d8ad73d96296ae503ee
->
https://github.com/oprypin/mkdocs/actions/runs/7076610215/job/19260039359

I have this:
```toml
[tool.hatch.envs.docs]
type = "pip-compile"
pip-compile-hashes = false
dependencies = [
   ...
]

[tool.hatch.envs.integration]
template = "docs"
```

I don't know why but it attempts to uninstall Hatch! And this fails only on Windows because only Windows cares that the process is trying to uninstall itself.

I actually realized one big problem with my setup that I just need to fix - for the `integration` env I intended to use the requirements file of the `docs` env, but actually it's still computed dynamically, so the lock file is missing. But why, if the lock file is missing, is it trying to uninstall Hatch?
